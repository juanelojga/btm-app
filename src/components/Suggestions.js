import React, {Component} from 'react'
import PropTypes from 'prop-types'

import './Suggestions.scss'

class Suggestions extends Component {
  constructor(props) {
    super(props)
    this.listRef = React.createRef()
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOutsideClick)
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOutsideClick)
  }

  handleClick = e => {
    this.props.onClick(e.target.id)
  }

  handleOutsideClick = e => {
    const node = this.listRef.current
    const {items, onClickOutside, isVisible} = this.props

    if (node && !node.contains(e.target) && items.length && isVisible) {
      onClickOutside()
    }
  }

  render() {
    const {items, isVisible} = this.props
    return (
      <div ref={this.listRef} className="list">
        {items.length && isVisible ? (
          <ul>
            {items.map(user => (
              <li key={user.id} id={user.id} onClick={this.handleClick}>
                {user.name}
              </li>
            ))}
          </ul>
        ) : (
          ''
        )}
      </div>
    )
  }
}

Suggestions.defaultProps = {
  items: [],
  isVisible: false,
}

Suggestions.propTypes = {
  items: PropTypes.array,
  onClick: PropTypes.func,
  onClickOutside: PropTypes.func,
  isVisible: PropTypes.bool,
}

export default Suggestions
