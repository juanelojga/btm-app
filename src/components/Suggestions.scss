.suggestion {
  width: 30rem;
  position: absolute;
  top: 3.5rem;
  border: 0.1rem solid #a0a0a1;
  background-color: rgba(#fff, 0.9);
  max-height: 15rem;
  overflow: auto;
  .list {
    list-style-type: none;
    margin: 0;
    padding: 0;
    li {
      font-size: 1.4rem;
      padding: 1rem;
      cursor: pointer;
    }
    li:hover {
      background: #eee;
      font-size: 1.6rem;
    }
  }
  .text {
    font-size: 1.4rem;
    padding: 1rem;
    cursor: pointer;
  }
}
