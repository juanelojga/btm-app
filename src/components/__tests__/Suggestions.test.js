import React from 'react'
import {render, fireEvent, cleanup} from 'react-testing-library'

import * as generators from '../../test/generators'
import Suggestions from '../Suggestions'

afterEach(cleanup)

test('calls onClick with the id of the user clicked', () => {
  const users = generators.generateUsers()

  const handleClick = jest.fn()
  const handleClickOutside = jest.fn()

  const {getByText} = render(
    <Suggestions items={users} isVisible={true} onClick={handleClick} onClickOutside={handleClickOutside} />,
  )

  fireEvent.click(getByText(users[0].name))

  expect(handleClick).toHaveBeenCalledTimes(1)
  expect(handleClick).toHaveBeenCalledWith(users[0].id)

  expect(handleClickOutside).not.toHaveBeenCalled()
})

test('snapshot', () => {
  const users = [{id: 1, name: 'Juan Almeida'}, {id: 2, name: 'Patricio Almeida'}]

  const {container} = render(<Suggestions items={users} isVisible={true} />)
  expect(container).toMatchSnapshot()
})
