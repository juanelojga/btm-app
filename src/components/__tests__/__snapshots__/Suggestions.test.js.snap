// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    class="suggestion"
  >
    <ul
      class="list"
    >
      <li
        id="1"
      >
        Juan Almeida
      </li>
      <li
        id="2"
      >
        Patricio Almeida
      </li>
    </ul>
  </div>
</div>
`;

exports[`snapshot loading 1`] = `
<div>
  <div
    class="suggestion"
  >
    <div
      class="loading show"
    >
      <div
        class="css-0"
      >
        <div
          class="css-1hpnt32-style"
        />
        <div
          class="css-vz2hb5-style"
        />
        <div
          class="css-1hpnt32-style"
        />
      </div>
    </div>
  </div>
</div>
`;
